### PR TITLE
Add some type annotations, most notably to `smbclient.open_file`

### PR DESCRIPTION
**Pull request by mon**
I won't lie, overloads are super verbose. But I worked off the ones for the stdlib `open()`, so it's "known good" at least.
I tested this by looking through all the test_smbclient_os opens and making sure the correct type was inferred.

I also added some extras so you can get typed connection details from an fd. It still kinda sucks (because something like the buffer sizes are None until `connect` is called) but at least it has intellisense.

Unfortunate API break: I couldn't work out a way to set the file/dir/pipe arg to specific values without forcing kwargs with * or moving the arg. I chose the latter, interested to hear your thoughts.

### CODE REVIEW

File: src/smbclient/_os.py
Comment (jborean93) [inline src/smbclient/_os.py:379]: I think we should hold off on explicitly documenting the `raw` value as our `SMBRawIO` type (and these subsequence `SMBBuffered*` types and just document the return values as the `io.Buffered*` types for now.

File: src/smbclient/_os.py
Comment (jborean93) [inline src/smbclient/_os.py:504]: So why does this need to be defined before the others? Is it because the overload only works based on the argument order?

File: src/smbclient/_os.py
Comment (mon) [reply src/smbclient/_os.py:379]: Ah, ok - I specifically added this because the tests access the raw FDs connection info.

Getting rid of this also lets me get rid of the tree/connection types in the other files causing the circular imports (or I can leave them in and gate with TYPE_CHECKING)

File: src/smbclient/_os.py
Comment (jborean93) [reply src/smbclient/_os.py:504]: I think let's just ignore the unbuffered file types and we can investigate a better API in the future without hopefully exposing a breaking change.

File: src/smbclient/_os.py
Comment (jborean93) [reply src/smbclient/_os.py:379]: Sounds like a win win :)

Comment (mon) [reply src/smbclient/_os.py:379]: I'll keep some of them (auth type, for example) but remove anything circular

**Comment by mon** [comment]
Hold up on a merge just for a moment, I'm getting a runtime error on py3.10 that the tests haven't caught, making sure I didn't break anything

**Comment by mon** [comment]
Alright, good to go once the tests finish running. Sorry for the churn!

**Review by jborean93 — APPROVED** [review]

**Comment by jborean93** [comment]
Thanks for working on this

**Comment by mon** [comment]
Thanks for the fast reviews, and having the most pleasant SMB library in the ecosystem :D

**Pull request merged**
